### PR TITLE
Best effort copying of `run` directory

### PR DESCRIPTION
**Pull request by michalpristas**
## What does this PR do?

This PR is using onErr hook provided by copy utility.
This onErr is raised whenever tool encounters an error, it's up to a caller to decide what to do with the error. In this PR we log error and return nil (swallowing the error) resulting in a copy process not being terminated. 

onErr is a custom hook introuced in my own fork of the tool (hence replace in `go.mod`) 
I also opened a PR with this hook (fully tested) in copy repo: https://github.com/otiai10/copy/pull/101
Once they decide to merge this we can remove our replace.

The other option was to perform copying by ourselves but the tool also handles symlinks and other edge cases already. 

Unit tests for open handles also present.

## Why is it important?

Prevents upgrade to fail on copy.

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [x] I have added an integration test or an E2E test

Fixes: #2433

### CODE REVIEW

**Comment by michalpristas** [comment]
need to replace fslock with something with better licensing

File: internal/pkg/agent/application/upgrade/upgrade_test.go [hunk 1]
```diff
@@ -9,15 +9,93 @@ import (
...
+				fl = flock.New(tc.From)
+				_, err := fl.TryLock()
+				require.NoError(t, err)
+			}
+
+			err := copyDir(l, tc.From, tc.To, tc.IgnoreErr)
+			require.Equal(t, tc.ExpectedErr, err != nil)
+			if fl != nil {
+				require.NoError(t, fl.Unlock())
+			}
```
Comment (ycombinator) [inline internal/pkg/agent/application/upgrade/upgrade_test.go:94]: Maybe invoke this via a `defer` so it will get executed even if the `require.Equal` on the previous line fails?

File: internal/pkg/agent/application/upgrade/upgrade.go [hunk 2]
```diff
@@ -343,11 +343,22 @@ func readDirs(dir string) ([]string, error) {
...
 }
 
-func copyDir(from, to string) error {
+func copyDir(l *logger.Logger, from, to string, ignoreErrs bool) error {
+	var onErr func(error) error
+
+	if ignoreErrs {
+		onErr = func(err error) error {
+			// ignore all errors, just log them
+			l.Debugf("ignoring error: failed to copy %q to %q: %s", from, to, err.Error())
```
Comment (cmacknz) [inline internal/pkg/agent/application/upgrade/upgrade.go:352]: I would consider making this an Info log, neither upgrades or errors are going to happen frequently enough for log spam to be an issue.

In general I would bias towards logging at the info level in the upgrade process so we aren't stuck trying to get rare failures to happen a second time with the log level changed to debug.

**Review by ycombinator — APPROVED** [review]
LGTM (pending @cmacknz's review feedback).

File: internal/pkg/agent/application/upgrade/upgrade.go (hunk 2 above)
Comment (jlind23) [reply internal/pkg/agent/application/upgrade/upgrade.go:352]: Could agree more, analysing debug logs is a pain for the upgrade process 😄 
Thanks @michalpristas for this btw
